coprocessor/fhevm-engine/fhevm-listener/src/health_check.rs

```diff
@@ -1,247 +1,108 @@
-use std::net::SocketAddr;
-use std::time::{SystemTime, UNIX_EPOCH};
-
 use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
 
-use alloy::providers::{ProviderBuilder, WsConnect};
+use tokio::sync::RwLock;
 
-use axum::{
-    extract::State,
-    http::StatusCode,
-    response::{IntoResponse, Json},
-    routing::get,
-    Router,
+use fhevm_engine_common::healthz_server::{
+    HealthCheckService, HealthStatus, Version,
 };
-use serde::Serialize;
-use sqlx::postgres::{PgConnectOptions, PgPoolOptions};
-use tokio::net::TcpListener;
-use tokio::sync::RwLock;
-use tokio_util::sync::CancellationToken;
-use tracing::{error, info};
+use fhevm_engine_common::types::BlockchainProvider;
+
+const IS_ALIVE_TICK_FRESHNESS: u64 = 20; // Not alive if tick is older
+const CONNECTED_TICK_FRESHNESS: u64 = 5; // Need to check connection if tick is older
 
 /// Represents the health status of the transaction sender service
-#[derive(Clone, Debug, Serialize)]
-pub struct Health {
-    pub status_code: u16,
-    pub status: &'static str,
-    /// Any error is unhealthy
-    pub healthy: bool,
-    /// Database connection status
-    pub database_connected: bool,
-    /// Blockchain provider connection status
-    pub blockchain_connected: bool,
-    pub message: String,
-    pub last_tick: u64,
+#[derive(Clone, Debug)]
+pub struct HealthCheck {
+    pub blockchain_timeout_tick: Tick,
+    pub blockchain_tick: Tick,
+    pub blockchain_provider: Arc<RwLock<Option<BlockchainProvider>>>,
+    pub database_pool: Arc<RwLock<sqlx::Pool<sqlx::Postgres>>>,
```

Comment:
It should be doable to avoid having RwLock around Pool.

As ref, the Pool is `Arc<PoolInner<DB>>` and in the SnS  workers we use it without the need to recreate it.  The pool can reestablish the connection as it runs a internal conn-retry.

Reply:
Yes, that's why we can clone them after accessing the lock. But the current database implem replace the pool when it needs to reconnect so it needs to be overwritten and with the current pool sharing in health check it implies a RWLock.